README.md
# Laravel Access Control List (ACL) System

## Installation
### Step 1 
Include via Composer

```bash
composer require gurinder/laravel-acl
```
### Step 2
Publish the migration, views(If required to customize), config(Optional)
``` bash
php artisan vendor:publish --tag="acl::migrations"

// Optional
php artisan vendor:publish --tag="acl::views"

// Optional
php artisan vendor:publish --tag="acl::config"
```
### Step 3
Run migration
```bash
php artisan migrate
```
## Config File
Configuration file information

```php
return [

    // Cache key to store acl data
    'cache_key' => 'gurinder.laravel-acl',
    
    // Cache Exiration time
    'cache_expiration_time' => 60 * 24,
    
    // Route prefix e.g. 'example.com/admin
    'route_prefix' => 'admin',

    // Route name as e.g use -> route('admin.roles.index')
    'route_as' => 'admin.',

    // Master roles - All new permission will be added to these roles automatically.
    'master_roles' => ['admin'],

    // Roles which are not editable
    'freezed_roles' => ['admin'],
    
    // Permissions which are not editable
    'freezed_permissions' => ['manage-acl', 'manage-users'],
    
    // Link back to dashboard or anywhere else
    'back_link' => [
        'label' => 'Dashboard',
        'url'   => '/home'
    ],
    
    // Search column on user models
    'user_search_columns' => [
        'id',
        'name',
        'email'
    ]

];
```

## Usage
Add `AclGuarded` Trait to your user model
```php
use Illuminate\Notifications\Notifiable;
use Illuminate\Foundation\Auth\User as Authenticatable;
use Gurinder\LaravelAcl\Traits\AclGuarded;

class User extends Authenticatable
{
    use Notifiable, AclGuarded;
    
    ...
}
```

Check if user has Role or Permission
```php
Gurinder\LaravelAcl\Package\Models\Role
Gurinder\LaravelAcl\Package\Models\Permission

$user->hasRole($role) // $role can be string, ID, or Role Model instance
$user->hasPermission($permission) // $permission can be string, ID, or Permission Model instance
```

You can also use permissions in blade directives
```php
@can('manage-acl')
    // The Current User Can Manage Acl
@elsecan('manage-users')
    // The Current User Can Manage Users
@endcan

// OR

@if (Auth::user()->can('manage-acl'))
    // The Current User Can Manage Acl
@endif

```

## Database Seeding
Make factories for roles and models
```php
use Faker\Generator as Faker;

$factory->define(\Gurinder\LaravelAcl\Package\Models\Role::class, function (Faker $faker) {
    return [
        'name' => $faker->word
    ];
});

$factory->define(\Gurinder\LaravelAcl\Package\Models\Permission::class, function (Faker $faker) {
    return [
        'name' => $faker->word
    ];
});
```
Make Seeder Class

```php
use Gurinder\LaravelAcl\Package\Models\Permission;
use Gurinder\LaravelAcl\Package\Models\Role;
use Illuminate\Database\Seeder;

class AccessControlListTableSeeder extends Seeder
{

    protected $roles = [ 'admin','editor'];

    protected $permissions = [
        'manage users',
        'manage acl',
        'manage posts'
    ];

    /**
     * Run the database seeds.
     *
     * @return void
     */
    public function run()
    {
        $roles = $this->seedRoles();
        $permissionsIds = $this->seedPermissions();
        $roles['admin']->syncPermissions($permissionsIds);
    }

    /**
     * @return array
     */
    protected function seedRoles()
    {
        $roles = [];
        
        foreach ($this->roles as $role) {
            $role = factory(Role::class)->create(['name' => $role, 'slug' => str_slug($role)]);
            $roles[$role->slug] = $role;
        }

        return $roles;
    }

    protected function seedPermissions()
    {
        $ids = [];

        foreach ($this->permissions as $label) {
            $ids[] = factory(Permission::class)->create(['name' => $label, 'slug' => str_slug($label)])->id;
        }

        return $ids;
    }
}
```

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

CHANGELOG.md
## 1.0.1 - 2018-06-07

### Added
- New Permission will be attached to master roles

## 1.0.0 - 2018-06-07

### Added
- Everything, initial release

src/Repositories/AclLedger.php
<?php

namespace Gurinder\LaravelAcl\Repositories;

use Illuminate\Support\Collection;
use Illuminate\Pagination\Paginator;
use Illuminate\Support\Facades\Cache;
use Gurinder\LaravelAcl\Package\Models\Role;
use Illuminate\Pagination\LengthAwarePaginator;
use Gurinder\LaravelAcl\Package\Models\Permission;
use Gurinder\LaravelAcl\Contracts\AclLedgerContract;

class AclLedger implements AclLedgerContract
{

    protected $usersAcl = [];

    /**
     * @param bool $paginate
     * @param int  $page
     * @param int  $perPage
     * @return Collection|LengthAwarePaginator|mixed
     */
    public function getRoles($paginate = false, $page = 1, $perPage = 15)
    {
        $roles = Cache::remember($this->getCacheKey('roles'), $this->getExpiration(), function () {
            return Role::orderBy('name')->with([
                'permissions' => function ($query) {
                    $query->orderBy('name');
                }
            ])->get();
        });

        return $paginate ? $this->paginate($roles, $page, $perPage, ['path' => request()->getBaseUrl()]) : $roles;

    }

    /**
     * @param bool $paginate
     * @param int  $page
     * @param int  $perPage
     * @return Collection|LengthAwarePaginator|mixed
     */
    public function getPermissions($paginate = false, $page = 1, $perPage = 15)
    {
        $permissions = Cache::remember($this->getCacheKey('permissions'), $this->getExpiration(), function () {
            return Permission::orderBy('name')->with([
                'roles' => function ($query) {
                    $query->orderBy('name');
                }
            ])->get();
        });

        return $paginate ? $this->paginate($permissions, $page, $perPage, ['path' => request()->getBaseUrl()]) : $permissions;

    }

    /**
     * @param $user
     * @return mixed
     */
    public function getUserAcl($user)
    {
        if (isset($this->usersAcl[$user->id])) {
            return $this->usersAcl[$user->id];
        }

        $acls = Cache::remember($this->getCacheKey('user', $user), $this->getExpiration(), function () use ($user) {
            return $user->rolesWithPermissions()->get();
        });

        $acls->each(function ($item, $key) use ($user) {

            $roleSlug = $item['slug'];
            $this->usersAcl[$user->id]['roles'][] = $roleSlug;

            collect($item['permissions'])->each(function ($item, $key) use ($user) {
                $permissionSlug = $item['slug'];
                $this->usersAcl[$user->id]['permissions'][] = $permissionSlug;
            });

        });

        return isset($this->usersAcl[$user->id]) ? $this->usersAcl[$user->id] : [];
    }

    /**
     * @param       $items
     * @param null  $page
     * @param int   $perPage
     * @param array $options
     * @return LengthAwarePaginator|mixed
     */
    public function paginate($items, $page = null, $perPage = 15, $options = [])
    {
        $page = $page ?: (Paginator::resolveCurrentPage() ?: 1);

        $items = $items instanceof Collection ? $items : Collection::make($items);

        return new LengthAwarePaginator($items->forPage($page, $perPage), $items->count(), $perPage, $page, $options);
    }

    /**
     * @param null $for
     * @param null $user
     * @return string
     */
    protected function getCacheKey($for = null, $user = null)
    {
        $configKey = config('acl.cache_key');

        $key = $for == 'user' ? "{$configKey}.acl.user.{$user->id}" : "{$configKey}.{$for}";

        return md5($key);
    }

    /**
     * @return float|\Illuminate\Config\Repository|int|mixed
     */
    protected function getExpiration()
    {
        return config('acl.cache_expiration_time') ?? 60 * 24;
    }

    /**
     * @return mixed|void
     */
    public function reset()
    {
        Cache::forget($this->getCacheKey('roles'));
        Cache::forget($this->getCacheKey('permissions'));
    }

    public function resetUserAcl($user)
    {
        Cache::forget($this->getCacheKey('user', $user));
    }

}